rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,518 @@
+    */
+    public class Product {
+     ...
+    }
+```
+
+Although this has already sneaked into some nix comments, this format is not considered best practice for various reasons.
+
+1. Essentially combines the cons of both worlds.
+2. It Takes up more space and needs autocompletion to fill missing `*` beginnings when extended.
+3. Starting every line with `*` creates visual conflicts with the markdown bullet list also starting with `*`.
+4. Pro: Indentation within is clear.
+5. Most nix users cannot identify with java or javascript and like predictable behavior.
+
+## Section headings considered
+
+### Markdown headings
+
+Example:
+
+```nix
+# somefile.nix
+
+    ## <Description or Tagline>
+    ## 
+    ## # Example
+    ##
+    ## <Comprehensive code>
+    ## 
+    ## # Type
+    ##    
+    ## <Type Signature>
+    mapAttrs = f: s: #...
+```
+
+| Pro | Con |
+|---|---|
+| Markdown is simple | doesn't visually distinguish from `##` starting the doc-string |
+| Using headings feels natural | Users may accidentally use those headings within natural language |
+| | Markdown recommends using newlines before and after headings which takes up a lot of vertical space |
+| | Markdown headings do not visually stand out from lines that already started with `##` |
+
+### Custom headings `@{Keyword}:`
+
+Example:
+
+```nix
+# somefile.nix
+
+    ## <Description or Tagline>
+    ## 
+    ## @Example:
+    ##
+    ## <Comprehensive code>
+    ## 
+    ## @Type:
+    ##    
+    ## <Type Signature>
+    mapAttrs = f: s: #...
+```
+
+| Pro | Con |
+|---|---|
+| Visually stands out | Is new syntax. Where Markdown could be more intuitive. Doc-strings already are Markdown. So why not use markdown |
+| Follows more closely the current convention |  |
+| Needs less vertical space | |
+| doesn't need newlines; everything could be even within a single line, allowing compression (may not be needed ?) | |
+
+## Alternative approach - just comments
+
+There is the idea from python that doc-strings are just strings, not even special ones. Comments will be docstrings if they follow specific placement rules. However, we thought this was a bad idea to follow. Such complex placement rules require the users to understand where those places are; with nix syntax, this is slightly more complex than with python. Because we don't have keywords such as `class MyClass():` or `def function():` where placement would be obvious
+
+# Detailed design
+[design]: #detailed-design
+
+## Proposed Solution (0,0) => `##` For docstring body and markdown headings `# H1`
+
+The following abstract rules describe how to write doc-strings.
+
+> I would be pleased if you comment about whether we should use `## {content}` or `/** {content} */`
+> I did write this RFC tailored towards `##` but using `/** */` is still an open discussion.
+
+### Format Rules (When choosing (0,0) )
+
+- [F100] - doc-string are all comments. That start with `##` or `#|` e.g. `## {content}`
```

Comment:
I'm confused by these (0, 0) and [F100] strings here, the meaning of these isn't explained anywhere.

Reply:
To me it looks like "I have no time to cleanup this copy-paste".

Reply:
> I'm confused by these (0, 0) and [F100] strings here, the meaning of these isn't explained anywhere.

Propably schould remove these. They dont make sense i think.

Reply:
@AndersonTorres please provide constructive feedback.